### DISCUSSION TITLE
How to transfer Typescript types to an edit-form component?

### DISCUSSION THREAD

**Discussion by zwl1619** [prisma/prisma#22789] [Q&A]
There is an edit-page, it gets data using prisma,as follow:

`app\dashboard\invoices\[id]\edit\page.tsx`
```
import Form from '@/app/ui/invoices/edit-form';
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();



async function fetchInvoiceById(id: string) {
  const data = await prisma.invoice.findUnique({
    where: {
      id: id,
    },
    select: {
      id: true,
      customerId: true,
      amount: true,
      status: true,
    },
  });
  const invoice = {
    ...data,
    amount: data!.amount / 100,   // Convert amount from cents to dollars
  };

  return invoice;
}



async function fetchCustomers() {
  const customers = await prisma.customer.findMany({
    select: {
      id: true,
      name: true,
    },
    orderBy: {
      name: 'asc',
    },
  });
  return customers;
}



export default async function Page({ params }: { params: { id: string } }) {
  const id = params.id;

  const [invoice, customers] = await Promise.all([
    fetchInvoiceById(id),
    fetchCustomers(),
  ]);

  return (
    <main>
      <Form invoice={invoice} customers={customers} />
    </main>
  );
}
```
the infered return type of function `fetchInvoiceById(id)` is like this:
[![enter image description here][1]][1]

and the infered return type of function `fetchCustomers()` is like this:
[![enter image description here][2]][2]

And the `Form` component is like this:   
it has some errors because the props `{invoice, customers}` has no types.    
`/app/ui/invoices/edit-form`:   
[![enter image description here][3]][3]

  
How to transfer the types into it from `app\dashboard\invoices\[id]\edit\page.tsx`?

  [1]: https://i.stack.imgur.com/ZkXXc.png
  [2]: https://i.stack.imgur.com/SlcPt.png
  [3]: https://i.stack.imgur.com/j2gsF.png

**Comment by ludralph** [comment]
Hi @zwl1619 👋 

Thank you for raising this question.

To transfer the types from your `fetchInvoiceById` and `fetchCustomers` functions to the Form component, you would typically define TypeScript interfaces or types for the expected shape of the data returned by these functions. Then, you would use these interfaces or types to type the props of the Form component.

In a shared TypeScript file accessible to both components (e.g., app/types.ts), define interfaces for the invoice and customer data types:
```ts
// app/types.ts
export interface Invoice {
  id: string;
  customerId: string;
  amount: number; // Assuming amount is now in dollars
  status: string;
}

export interface Customer {
  id: string;
  name: string;
}
```
Import the interfaces in the Form component.
```tsx
// app/ui/invoices/edit-form.tsx
import { Invoice, Customer } from '@/app/types';

interface FormProps {
  invoice: Invoice | undefined;
  customers: Customer[];
}

const Form: React.FC<FormProps> = ({ invoice, customers }) => {
  // ... rest of your component code
};
```

In the Page component, pass the props with the correct types
```tsx
// app\dashboard\invoices\[id]\edit\page.tsx
// ... other imports
import { Form } from '@/app/ui/invoices/edit-form';

// ... other code
return (
  <main>
    <Form invoice={invoice} customers={customers} />
  </main>
);
```

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
